chore: Make keyring non-interactive

fredcarle opened the pull request:
## Relevant issue(s)

Resolves #2995 

## Description

This PR makes the keyring interactions non-interactive by requiring the keyring password to be set as an environment variable secret. It also adds support for that secret to be stored in a `.env` file in the working directory or in a file at a path defined by the `--secret-file` flag.

Making the keyring non-interactive is necessary to support automated deployments. 

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

make test

Also tested manually starting Defra with a `.env` file to make sure that the secret was picked up.

Specify the platform(s) on which this was tested:
- MacOS

nasdf commented on cli/config.go:
```diff
@@ -94,6 +96,9 @@ var configDefaults = map[string]any{
 // defaultConfig returns a new config with default values.
 func defaultConfig() *viper.Viper {
+	// load environment variables from .env file if one exists
+	_ = godotenv.Load()
```
todo: viper has built in support for envfiles. We should be able to remove this dependency

fredcarle replied:
I think viper can read from only one config file at a time. We already have the defra config file that is used with viper.

nasdf replied:
ahh okay that seems like an odd choice on their part

shahzadlone reviewed (commented):
might want to also ignore the `.env` file in the `.gitignore`

shahzadlone reviewed (commented):
I like the approach, just had some questions:

question: Where is the loading of the environment variables from '.env' file being done?

question: When you say `root` directory what do you mean i.e. what are the paths that this `.env` file will be picked from (and the order of priority if many paths)?

question: Have you tested this works when `.env` flow?

islamaliev commented on cli/keyring_export_test.go:
```diff
@@ -13,21 +13,20 @@ package cli
-	readPassword = func(_ *cobra.Command, _ string) ([]byte, error) {
-		return []byte("secret"), nil
-	}
+	err := os.Setenv("DEFRA_KEYRING_SECRET", "password")
```
question: don't we want to keep at least one test that uses the old prompt?

fredcarle replied:
The prompt has been removed entirely.

fredcarle commented:
> question: Where is the loading of the environment variables from '.env' file being done?

@shahzadlone, it's done from where Defra is started.

> question: When you say root directory what do you mean i.e. what are the paths that this .env file will be picked from (and the order of priority if many paths)?

Currently there is only one path and it's `./.env`. Eventually I would like to add a flag to let users set the path.

> question: Have you tested this works when .env flow?

Yes I've tested this manually.

fredcarle commented:
> > question: When you say root directory what do you mean i.e. what are the paths that this .env file will be picked from (and the order of priority if many paths)?
> 
> Currently there is only one path and it's `./.env`. Eventually I would like to add a flag to let users set the path.

@shahzadlone I've added a flag to let the user define the path and updated the documentation to say `working directory` instead of `root directory`.

nasdf commented on cli/keyring_generate.go:
thought: I'm not sure how useful this command is now that we generate keys on start

fredcarle replied:
It sensible if someone wants to generate the keys first for some reason. I had initially removed it then decided to keep it as a convenience.

nasdf commented on cli/config.go:
```diff
@@ -42,29 +43,31 @@ var configPaths = []string{
+	"keyring-path":        "keyring.path",
+	"no-keyring":          "keyring.disabled",
+	"no-encryption-key":   "keyring.noencryptionkey",
+	"keyring-secret-file": "keyring.secretfile",
```
question: What do you think about making this more general purpose and renaming to `secretfile`? If we decide to add more secrets in the future it might be confusing having the keyring name.

fredcarle replied:
I did this initially and then reverted. I think I'm more and more in favour of making it general purpose. I'll change it.

fredcarle replied:
Moved to `secretfile`.

nasdf reviewed (approved):
A few thoughts and questions but nothing blocking.

shahzadlone reviewed (approved):
LGTM

fredcarle commented on cli/config.go:
```diff
@@ -42,29 +43,31 @@ var configPaths = []string{
+	"keyring-backend":     "keyring.backend",
+	"keyring-path":        "keyring.path",
+	"no-keyring":          "keyring.disabled",
+	"no-encryption-key":   "keyring.noencryptionkey",
```
That's a sensible thought. I'll look into it.

fredcarle replied:
changed to `datastore.noencryption` for now. We can add a keyname configuration later.

fredcarle commented on cli/config.go:
```diff
@@ -159,6 +163,9 @@ func loadConfig(rootdir string, flags *pflag.FlagSet) (*viper.Viper, error) {
 	}
 
+	// load environment variables from .env file if one exists
+	_ = godotenv.Load(cfg.GetString("keyring.secretfile"))
```
Added error handling when it's not an file not found error.

fredcarle replied:
The `.env` file is optional so I don't want to pollute the logs with statements of a missing optional file.
